components/core/src/clp/streaming_archive/single_file_archive/writer.cpp

```diff
@@ -0,0 +1,224 @@
+#include <spdlog.h>
+
+#include "../../Defs.h"
+#include "../../ErrorCode.hpp"
+#include "../../FileReader.hpp"
+#include "../../FileWriter.hpp"
+#include "../../TraceableException.hpp"
+#include "../ArchiveMetadata.hpp"
+#include "../Constants.hpp"
+#include "Defs.hpp"
+
+namespace clp::streaming_archive::single_file_archive {
+
+namespace {
+constexpr size_t cReadBlockSize = 4096;
+
+/**
+ * Generates metadata for the file section of a single-file archive. The metadata consists
+ * of a list of file names and their corresponding starting offsets.
+ *
+ * @param multi_file_archive_path
+ * @param num_segments
+ * @return Vector containing a `FileInfo` struct for every file in the multi-file archive.
+ * @throws `std::filesystem::filesystem_error` if `stat` on archive file fails.
+ */
+[[nodiscard]] auto
+get_archive_file_infos(std::filesystem::path const& multi_file_archive_path, size_t num_segments)
+        -> std::vector<FileInfo>;
+
+/**
+ * Combines file section metadata and the number of segments into single-file archive metadata,
+ * then serializes the metadata into MsgPack format.
+ *
+ * @param multi_file_archive_path
+ * @param num_segments
+ * @return Packed metadata.
+ */
+[[nodiscard]] auto pack_single_file_archive_metadata(
+        std::filesystem::path const& multi_file_archive_path,
+        size_t num_segments
+) -> std::stringstream;
+
+/**
+ * Writes single-file archive header.
+ *
+ * @param single_file_archive_writer
+ * @param packed_metadata_size
+ */
+auto write_archive_header(FileWriter& single_file_archive_writer, size_t packed_metadata_size)
+        -> void;
+
+/**
+ * Writes packed single-file archive metadata.
+ *
+ * @param single_file_archive_writer
+ * @param packed_metadata Packed metadata.
+ */
+auto write_packed_archive_metadata(
+        FileWriter& single_file_archive_writer,
+        std::stringstream const& packed_metadata
+) -> void;
+
+/**
+ * Reads the content of a file and writes it to the single-file archive.
+ * @param file_path
+ * @param single_file_archive_writer
+ * @throws OperationFailed if reading the file fails.
+ */
+auto
+write_archive_file(std::filesystem::path const& file_path, FileWriter& single_file_archive_writer)
+        -> void;
+
+/**
+ * Iterates over files in the multi-file archive and copies their contents to the single-file
+ * archive.
+ *
+ * @param single_file_archive_writer
+ * @param multi_file_archive_path
+ * @param num_segments
+ */
+auto write_archive_files(
+        FileWriter& single_file_archive_writer,
+        std::filesystem::path const& multi_file_archive_path,
+        size_t num_segments
+) -> void;
+
+auto
+get_archive_file_infos(std::filesystem::path const& multi_file_archive_path, size_t num_segments)
+        -> std::vector<FileInfo> {
+    std::vector<FileInfo> files;
+    uint64_t offset{};
+
+    for (auto const& static_archive_file_name : cStaticArchiveFileNames) {
+        files.emplace_back(FileInfo{std::string(static_archive_file_name), offset});
```

Comment:
```suggestion
        files.emplace_back(std::string{static_archive_file_name}, offset);
```

`emplace_back` forwards the parameters to the object's constructor, right?